scripts/lb-wrapper

```diff
@@ -1,48 +1,50 @@
 #!/bin/bash
 
 LOG_FILE="/var/log/xklb.log"
+LOG_LEVEL="-vv"
 XKLB_EXECUTABLE="${XKLB_EXECUTABLE:-lb}"
 TMP_DOWNLOADS_DIR="/library/downloads/calibre-web"
 SURVEY_DB_FILE="${TMP_DOWNLOADS_DIR}/survey.db"
 URL="$1"
 FORMAT_OPTIONS="--format best --format-sort 'tbr~1000'"
-XKLB_FULL_CMD="${XKLB_EXECUTABLE} tubeadd ${SURVEY_DB_FILE} ${URL} --verbose && ${XKLB_EXECUTABLE} dl ${SURVEY_DB_FILE} --prefix ${TMP_DOWNLOADS_DIR} --write-thumbnail ${FORMAT_OPTIONS} --video ${URL} --verbose"
+XKLB_FULL_CMD="${XKLB_EXECUTABLE} tubeadd ${SURVEY_DB_FILE} ${URL} ${LOG_LEVEL}&& ${XKLB_EXECUTABLE} dl ${SURVEY_DB_FILE} --prefix ${TMP_DOWNLOADS_DIR} --write-thumbnail ${FORMAT_OPTIONS} --video ${URL} ${LOG_LEVEL}"
 
 mkdir -p ${TMP_DOWNLOADS_DIR}
 
-# Function to log messages (and errors especially!)
+# Function to log messages with different levels
 log() {
-    echo "$(date +'%Y-%m-%d %H:%M:%S') - $1" | tee -a ${LOG_FILE}
+    local level=$1
+    local message=$2
+    echo "$(date +'%Y-%m-%d %H:%M:%S') - [$level] $message" >> ${LOG_FILE}
 }
 
 if ! command -v "${XKLB_EXECUTABLE}"; then
-    log "xklb could not be found. Please install xklb and try again."
+    log "Error" "xklb could not be found. Please install xklb and try again."
     exit 1
 fi
 
 if [ $# -eq 0 ]; then
-    log "No arguments provided. Please provide a URL to download."
+    log "Error" "No arguments provided. Please provide a URL to download."
     exit 1
 fi
 
-# URL validation already taken care of by cps/static/js/main.js Lines 194-195
-# which (1) trims outer whitespace, and (2) prepends https:// if URL doesn't
-# already begin with http:// or https://
-# (Test below meant well, enforcing much the same things with logging, but
-# let's avoid duplicate code insofar as possible!)
-# if [[ ! ${URL} =~ ^http[s]?:// ]]; then
-#     log 'Invalid URL: xklb commands require URLs begin with "http://" or "https://"'
-#     exit 1
-# fi
-
 if mv ${SURVEY_DB_FILE} ${SURVEY_DB_FILE}.$(date +%F_%T_%Z) 2> /dev/null; then
-    log "Old ${SURVEY_DB_FILE} moved aside."
+    log "Info" "Old ${SURVEY_DB_FILE} moved aside."
 fi
 
-log "Running command: ${XKLB_FULL_CMD}"
-if OUTPUT=$(eval "${XKLB_FULL_CMD}"); then
-    log "Download completed successfully."
+log "Info" "Running command: ${XKLB_FULL_CMD}"
+
+# Run the command in the background and capture the PID
+eval "${XKLB_FULL_CMD}" > >(while read -r line; do log "Info" "$line"; done) 2> >(while read -r line; do log "Error" "$line"; done) &
+pid=$!
+
+# Wait for the background process to complete
+wait $pid
+
+# Check the exit status of the command
+if [ $? -eq 0 ]; then
```

Comment:
If the return code from `wait $pid` is in fact meaningful:

- Should it be passed along to the `log` function, i.e. log file and/or journalctl, or whatever evolves here?

- Do any of the 20+ questions/answers at https://mywiki.wooledge.org/ProcessManagement reasonably describe the overall `wait` code mechanism/pattern here, to help others understand?

Reply:
What's the reason to background the xklb command on Line 38?

(Conversely, if there's no good reason for the `&` on the right side of Line 38, then also remove Lines 39 and 42?)

Reply:
> If the return code from `wait $pid` is in fact meaningful:
> 
>     * Should it be passed along to the `log` function, i.e. log file and/or journalctl, or whatever evolves here?
> 
>     * Do any of the 20+ questions/answers at https://mywiki.wooledge.org/ProcessManagement reasonably describe the overall `wait` code mechanism/pattern here, to help others understand?

It is. We have in fact 2 stdout stream to handle:
 1. The xklb subprocess (lb-wrapper)
 2. the yt-dlp[sub]subprocess, which xklb runs in the background.

We need to read lines and wait for the yt-dlp subprocess to finish. The only problem is it doesnt print to stdout, which will be corrected in the next commit.

Reply:
> What's the reason to background the xklb command on Line 38?
> 
> (Conversely, if there's no good reason for the `&` on the right side of Line 38, then also remove Lines 39 and 42?)

I don't think I understand very well your assumptions. When lb-wrapper is run, calibre-web is expecting a response from xklb execution. To avoid blocking, the whole script is run in a thread. In the background xklb do database operations and fires yt-dlp, b-wrapper is also expecting a response from yt-dlp.

Reply:
> The only problem is it doesnt print to stdout

What exactly doesn't print to STDOUT and why?

> which will be corrected in the next commit.

Link please?

Reply:
> > The only problem is it doesnt print to stdout
> 
> What exactly doesn't print to STDOUT and why?
> 
> > which will be corrected in the next commit.
> 
> Link please?

Lb-wrapper is a command. It returns 0 and 1. It also prints to STDOUT everything you echo in the script. Additionally, I redirected yt-dlp download progress information to its STDOUT so `cps/tasks/download.py`
can read it in real time for polling purposes.

Reply:
> What exactly doesn't print to STDOUT and why?

The STDOUT in question is lb-wrapper's. What didn't print to it is yt-dlp download progress information. This over growing PR was built for this, to capture it efficiently.

It didn't print nothing because the code was buggy. I had to try with a temporary polling file until I got to adopt redirecting to STDOUT what I need from yt-dlp.

> Link please?

Reply:
Thank you for shedding light:

Helpfully hinting at what is being attempted (and why!)